Class org.languagetool.GlobalConfig not found on ArchLinux

Hi!

I just installed the LanguageTool Linter extension for VSCode on ArchLinux. I choose to let the extension manage the LanguageTool server.

I use the following configuration in `settings.json`:

```json
"languageToolLinter.enabled": true,
"languageToolLinter.serviceType": "managed",
"languageToolLinter.managed.jarFile": "/usr/share/java/languagetool/languagetool-server.jar",
"languageToolLinter.languageTool.motherTongue": "fr",
"languageToolLinter.languageTool.language": "en-US",
```

The extension output in VSCode is shown below:

```
Starting managed service.
Picked up JAVA_TOOL_OPTIONS:  -Djdk.gtk.version=3

LanguageTool Linter Activated!
Exception in thread "main" java.lang.NoClassDefFoundError: org/languagetool/GlobalConfig
	at org.languagetool.server.HTTPServerConfig.<init>(HTTPServerConfig.java:94)
	at org.languagetool.server.HTTPServer.main(HTTPServer.java:140)
Caused by: java.lang.ClassNotFoundException: org.languagetool.GlobalConfig
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:604)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
	... 2 more

Managed service command failed: java -cp /usr/share/java/languagetool/languagetool-server.jar org.languagetool.server.HTTPServer --port 8000
Error Message: Command failed with exit code 1: java -cp /usr/share/java/languagetool/languagetool-server.jar org.languagetool.server.HTTPServer --port 8000
```

After looking at the LanguageTool launch script available on ArchLinux, I think the _classpath_ is not well defined when this extension run the server:

```shell
#!/bin/bash

file_present=false
asked_for_help=false
start_server=false
http_server=false

declare -a argv=("$@")

i=0
for arg in "$@"; do
  if [[ $arg != -* ]]; then
    file_present=true
  fi
  if [[ $arg == --config* ]]; then
    start_server=true
  fi
  if [[ $arg == --http ]]; then
    start_server=true
    http_server=true
    unset argv[$i]
  fi
  if [[ $arg == -h ]]; then
    asked_for_help=true
    unset argv[$i]
  fi
  if [[ $arg == --help ]]; then
    asked_for_help=true
    unset argv[$i]
  fi
  ((i++))
done

CP=/usr/share/languagetool
for name in /usr/share/java/languagetool/*.jar ; do
  CP=$CP:$name
done

CLI_command='org.languagetool.commandline.Main'
GUI_command='org.languagetool.gui.Main'
NSRV_command='org.languagetool.server.HTTPServer'
SSRV_command='org.languagetool.server.HTTPSServer'

if $asked_for_help; then
  echo "Command-line interface (CLI) help:"
  "$JAVA_HOME/bin/java" -cp $CP $CLI_command -h | sed "s/java -jar languagetool-commandline.jar/languagetool/"
  echo
  echo "Graphical user interface (GUI) help:"
  "$JAVA_HOME/bin/java" -cp $CP $GUI_command -h | sed "s/java org.languagetool.gui.Main/languagetool/"
  echo
  echo "HTTP server help:"
  "$JAVA_HOME/bin/java" -cp $CP $NSRV_command -h | sed "s/HTTPServer/languagetool --http/"
  echo
  echo "HTTPS server help:"
  "$JAVA_HOME/bin/java" -cp $CP $SSRV_command -h | sed "s/HTTPSServer/languagetool/"
else
  if $start_server; then
    if $http_server; then
      "$JAVA_HOME/bin/java" -cp $CP $NSRV_command "${argv[@]}"
    else
      "$JAVA_HOME/bin/java" -cp $CP $SSRV_command "${argv[@]}"
    fi
  else
    if $file_present; then
      "$JAVA_HOME/bin/java" -cp $CP $CLI_command "${argv[@]}"
    else
      "$JAVA_HOME/bin/java" -cp $CP $GUI_command "${argv[@]}"
    fi
  fi
fi
```

Each JAR file located in the `/usr/share/java/languagetool` directory is appended to the _classpath_. As your extension only include the the `languagetool-server.jar` file to the classpath, some classes are not found.

I don't know how LanguageTool is packaged on other Linux distribution, but it seems your extension does not work well with the ArchLinux package.

I think about another solution: it may be useful to specify a command to launch instead of a path to the JAR file.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,6 +4,14 @@ All notable changes to the "languagetool-linter" extension will be documented in
 
 ## [Unreleased]
 
+### Added
+
+* `managed.classPath` setting supports multiple paths and file globbing via [node-glob](https://github.com/isaacs/node-glob). This accomodates various install methods. For example, on Arch Linux using the pacman LanguageTool package, you would set this to `/usr/share/java/languagetool/*.jar`
+
+### Deprecated
+
+* Deprecated `managed.jarFile` setting in favor of `managed.classPath` to align with how the setting was used.
+
 ## [0.3.2] - 2019-11-04
 
 ### Fixed

--- a/package.json
+++ b/package.json
@@ -101,7 +101,7 @@
         "languageToolLinter.managed.classPath": {
           "type": "string",
           "default": "",
-          "description": "Custom classpath for managed server. Use filesystem separator. File globbing per node-glob."
+          "markdownDescription": "Custom classpath for managed server. Separate multiple paths using the native filesystem separator. Supports file globbing via [node-glob](https://github.com/isaacs/node-glob)."
         },
         "languageToolLinter.languageTool.language": {
           "type": "string",

--- a/package.json
+++ b/package.json
@@ -98,6 +98,11 @@
           "default": "",
           "description": "Path to languagetool-server.jar on your local machine."
         },
+        "languageToolLinter.managed.classPath": {
+          "type": "string",
+          "default": "",
+          "description": "Custom classpath for managed server. Use filesystem separator. File globbing per node-glob."
+        },
         "languageToolLinter.languageTool.language": {
           "type": "string",
           "default": "auto",
@@ -322,7 +327,7 @@
     "@types/mocha": "^5.2.6",
     "@types/node": "^10.17.4",
     "@types/vscode": "^1.39.0",
-    "glob": "^7.1.5",
+    "glob": "^7.1.6",
     "mocha": "^6.2.2",
     "tslint": "^5.20.1",
     "typescript": "^3.6.4",

--- a/src/common/configuration.ts
+++ b/src/common/configuration.ts
@@ -117,20 +117,16 @@ export class ConfigurationManager implements Disposable {
       let jarFile: string = this.get("managed.jarFile") as string;
       let classPath: string = this.get("managed.classPath") as string;
       let classPathFiles: string[] = [];
-      let classPathString: string = "";
       if (jarFile !== "") {
-        classPathString = jarFile;
+        classPathFiles.push(jarFile);
       }
       classPath.split(path.delimiter).forEach((globPattern) => {
         console.log("Glob: " + globPattern);
-        glob(globPattern, (error, matches) => {
-          console.log("Glob Matches: " + matches);
-          matches.forEach((match) => {
-            classPathString += path.delimiter + match;
-            console.log("Class Path String: " + classPathString);
-          });
+        glob.sync(globPattern).forEach((match) => {
+          classPathFiles.push(match);
         });
       });
+      let classPathString: string = classPathFiles.join(path.delimiter);
       console.log("class path: " + classPathString);
       this.stopManagedService();
       portfinder.getPort({ host: "127.0.0.1" }, (error, port) => {

--- a/src/common/configuration.ts
+++ b/src/common/configuration.ts
@@ -1,4 +1,4 @@
-import { TextDocument, WorkspaceConfiguration, workspace, ConfigurationChangeEvent, Disposable } from 'vscode';
+import { TextDocument, WorkspaceConfiguration, workspace, ConfigurationChangeEvent, Disposable, window } from 'vscode';
 import { LT_DOCUMENT_LANGUAGE_IDS, LT_CONFIGURATION_ROOT, LT_SERVICE_PARAMETERS, LT_SERVICE_EXTERNAL, LT_CHECK_PATH, LT_SERVICE_MANAGED, LT_SERVICE_PUBLIC, LT_PUBLIC_URL, LT_OUTPUT_CHANNEL } from './constants';
 import * as portfinder from 'portfinder';
 import * as execa from "execa";
@@ -37,7 +37,7 @@ export class ConfigurationManager implements Disposable {
         (event.affectsConfiguration("languageToolLinter.managed.jarFile")
         || event.affectsConfiguration("languageToolLinter.managed.classPath"))
       ) {
-      this.restartManagedService();
+      this.startManagedService();
     }
   }
 
@@ -117,17 +117,19 @@ export class ConfigurationManager implements Disposable {
       let jarFile: string = this.get("managed.jarFile") as string;
       let classPath: string = this.get("managed.classPath") as string;
       let classPathFiles: string[] = [];
+      // DEPRECATED
       if (jarFile !== "") {
+        window.showWarningMessage('"LanguageTool Linter > Managed: Jar File" is deprecated. Please use "LanguageTool > Managed: Class Path" instead.');
         classPathFiles.push(jarFile);
       }
-      classPath.split(path.delimiter).forEach((globPattern) => {
-        console.log("Glob: " + globPattern);
-        glob.sync(globPattern).forEach((match) => {
-          classPathFiles.push(match);
+      if (classPath !== "") {
+        classPath.split(path.delimiter).forEach((globPattern) => {
+          glob.sync(globPattern).forEach((match) => {
+            classPathFiles.push(match);
+          });
         });
-      });
+      }
       let classPathString: string = classPathFiles.join(path.delimiter);
-      console.log("class path: " + classPathString);
       this.stopManagedService();
       portfinder.getPort({ host: "127.0.0.1" }, (error, port) => {
         if (error) {
@@ -172,11 +174,6 @@ export class ConfigurationManager implements Disposable {
       this.process.cancel();
       this.process = undefined;
     }
-  }
-
-  restartManagedService(): void {
-    this.stopManagedService();
-    this.startManagedService();
   }
 
 }

--- a/src/common/configuration.ts
+++ b/src/common/configuration.ts
@@ -2,6 +2,8 @@ import { TextDocument, WorkspaceConfiguration, workspace, ConfigurationChangeEve
 import { LT_DOCUMENT_LANGUAGE_IDS, LT_CONFIGURATION_ROOT, LT_SERVICE_PARAMETERS, LT_SERVICE_EXTERNAL, LT_CHECK_PATH, LT_SERVICE_MANAGED, LT_SERVICE_PUBLIC, LT_PUBLIC_URL, LT_OUTPUT_CHANNEL } from './constants';
 import * as portfinder from 'portfinder';
 import * as execa from "execa";
+import * as path from "path";
+import * as glob from "glob";
 
 export class ConfigurationManager implements Disposable {
   private config: WorkspaceConfiguration;
@@ -31,6 +33,11 @@ export class ConfigurationManager implements Disposable {
           this.stopManagedService();
           break;
       }
+    } else if (this.getServiceType() === LT_SERVICE_MANAGED &&
+        (event.affectsConfiguration("languageToolLinter.managed.jarFile")
+        || event.affectsConfiguration("languageToolLinter.managed.classPath"))
+      ) {
+      this.restartManagedService();
     }
   }
 
@@ -108,6 +115,23 @@ export class ConfigurationManager implements Disposable {
   private startManagedService(): void {
     if (this.getServiceType() === LT_SERVICE_MANAGED) {
       let jarFile: string = this.get("managed.jarFile") as string;
+      let classPath: string = this.get("managed.classPath") as string;
+      let classPathFiles: string[] = [];
+      let classPathString: string = "";
+      if (jarFile !== "") {
+        classPathString = jarFile;
+      }
+      classPath.split(path.delimiter).forEach((globPattern) => {
+        console.log("Glob: " + globPattern);
+        glob(globPattern, (error, matches) => {
+          console.log("Glob Matches: " + matches);
+          matches.forEach((match) => {
+            classPathString += path.delimiter + match;
+            console.log("Class Path String: " + classPathString);
+          });
+        });
+      });
+      console.log("class path: " + classPathString);
       this.stopManagedService();
       portfinder.getPort({ host: "127.0.0.1" }, (error, port) => {
         if (error) {
@@ -117,7 +141,7 @@ export class ConfigurationManager implements Disposable {
           this.setManagedServicePort(port);
           let args: string[] = [
             "-cp",
-            jarFile,
+            classPathString,
             "org.languagetool.server.HTTPServer",
             "--port",
             port.toString()
@@ -154,5 +178,9 @@ export class ConfigurationManager implements Disposable {
     }
   }
 
+  restartManagedService(): void {
+    this.stopManagedService();
+    this.startManagedService();
+  }
 
 }

